### PR TITLE
Audio synchronisation improvements

### PR DESCRIPTION
**Pull request by bylaws**
- Should help to prevent stalls from needing to happen by attempting to tie the guest sample tracking to the host
- I am still unsure about the second patch, needs checking against RE to see if this wait time is included in the execution time accounting. The implementation is also kinda dodgy, ideally you'd keep track of reserved and submitted buffers and ensure they match

### CODE REVIEW

**Comment by gidoly** [comment]
after testing it produced 2 issues.

1. audout games performance decrease
tested games are:
super mario sunshine. 40fps > 28fps
N64 NSO emulater. 60fps > 45fps
non aud out games seems remain good.
2. freeze while closing game.
yuzu will freeze after click ' stop emulation button ' and never recover from it

File: src/audio_core/sink/sink_stream.cpp [hunk 1]
```diff
@@ -282,4 +290,17 @@ void SinkStream::Unstall() {
...
+    std::scoped_lock lk{sample_count_lock};
+    auto time_delta{std::chrono::system_clock::now() - last_sample_count_update_time};
+    auto exp_played_sample_count{min_played_sample_count +
+                                 TargetSampleRate * (time_delta / std::chrono::seconds{1})};
+    return std::min<u64>(exp_played_sample_count, max_played_sample_count);
+}
+
+void SinkStream::WaitFreeSpace() {
+    while (queued_buffers >= max_queue_size)
+        release_event.Wait();
```
Comment (Morph1984) [inline src/audio_core/sink/sink_stream.cpp:303]:
```suggestion
    while (queued_buffers >= max_queue_size) {
        release_event.Wait();
    }
```

File: src/audio_core/sink/sink_stream.cpp [hunk 2]
```diff
@@ -260,26 +247,31 @@ void SinkStream::ProcessAudioOutAndRender(std::span<s16> output_buffer, std::siz
...
 
-void SinkStream::Unstall() {
-    std::scoped_lock lk{stall_guard};
-    if (!stalled_lock) {
-        return;
+void SinkStream::WaitFreeSpace(std::stop_token stop_token) {
+    while (queued_buffers >= max_queue_size) {
+        std::unique_lock lk{release_mutex};
+        Common::CondvarWait(release_cv, lk, stop_token, [this]() { return release_signalled; });
+        release_signalled = false;
```
Comment (liamwhite) [inline src/audio_core/sink/sink_stream.cpp:273]: Do you really need `release_signalled`? Couldn't you just wait for the underlying condition to be true, i.e.
```suggestion
    Common::CondvarWait(release_cv, lk, stop_token, [this]() { return queued_buffers < max_queue_size; });
```

And then in ProcessAudioOutAndRender, after dequeuing a buffer,
```cpp
            {
                std::unique_lock lk{release_mutex};
            }
            release_cv.notify_one();
```

**Comment by legend800** [comment]
Not sure if it was this audio change but within the past week or two, TMNT Shredder's Revenge finally has correct audio when using single-core, which is the only way to boot the game. Before, it was staticy, laggy and cutoff.

**Review by Kelebek1 — APPROVED** [review]

**Review by bunnei — APPROVED** [review]

**Comment by octopushugger** [comment]
Not 100% certain but this PR seems like the most likely culprit for the broken audio in Crystal Chronicles Remastered. Nonstop popping in all audio playback throughout the game now. I rolled back to build 1388 where the audio sounds fine.

**Comment by tiagocoelho231** [comment]
Rune Factory 3 Special audio also has nonstop popping since version 1389. It was working fine in 1388.

**Comment by german77** [comment]
We are aware of the issues, a fix has already been made we are just waiting for the author to open a PR with the fix

**Comment by HurricanePootis** [comment]
This patch is causing Minecraft Dungeons to crash after studio intro videos are done https://github.com/yuzu-emu/yuzu/issues/10029

**Comment by jroberts2010** [comment]
Popping can be heard in Fire Emblem: Three Houses starting from this PR. Starts immediately in title menu.

**Pull request merged**